src/MeshKernel/CurvilinearGrid/CurvilinearGrid.cpp

```diff
@@ -506,91 +508,134 @@ void CurvilinearGrid::InsertFace(Point const& point)
-                                first.m_n < m_gridNodes[0].size() &&
-                                second.m_m < m_gridNodes.size() &&
-                                second.m_n < m_gridNodes[0].size();
-        return !validIndex || m_gridNodes[first.m_m][first.m_n].IsValid() && !m_gridNodes[second.m_m][second.m_n].IsValid();
-    };
+    // If both nodes are invalid, we can substitute the invalid values. New allocation is not needed.
+    bool const areNodesValid = m_gridNodes[firstNode.m_m][firstNode.m_n].IsValid() && m_gridNodes[secondNode.m_m][secondNode.m_n].IsValid();
 
     // Allocation depends on directions
-    if (m_gridNodesTypes[firstNode.m_m][firstNode.m_n] == NodeType::Left ||
-        m_gridNodesTypes[secondNode.m_m][secondNode.m_n] == NodeType::Left)
+    bool gridSizeChanged = false;
+    auto const gridLineType = GetBoundaryGridLineType(firstNode, secondNode);
+    if (gridLineType == BoundaryGridLineType::Left && areNodesValid)
     {
-        auto const firstNewNodeCoordinates = m_gridNodes[firstNode.m_m][firstNode.m_n] * 2.0 - m_gridNodes[firstNode.m_m + 1][firstNode.m_n];
-        auto const secondNewNodeCoordinates = m_gridNodes[secondNode.m_m][secondNode.m_n] * 2.0 - m_gridNodes[secondNode.m_m + 1][secondNode.m_n];
-        if (isAllocationNeeded({firstNode.m_m - 1, firstNode.m_n}, {secondNode.m_m - 1, secondNode.m_n}))
+        m_gridNodes.emplace(m_gridNodes.begin(), std::vector<Point>(m_gridNodes[0].size()));
+        gridSizeChanged = true;
+    }
+    if (gridLineType == BoundaryGridLineType::Right && areNodesValid)
+    {
+        m_gridNodes.emplace_back(std::vector<Point>(m_gridNodes[0].size()));
+        gridSizeChanged = true;
+    }
+    if (gridLineType == BoundaryGridLineType::Up && areNodesValid)
+    {
+        for (auto& gridNodes : m_gridNodes)
         {
-            m_gridNodes.emplace(m_gridNodes.begin(), std::vector<Point>(m_gridNodes[0].size()));
-            // Assign the new coordinates
-            m_gridNodes[firstNode.m_m][firstNode.m_n] = firstNewNodeCoordinates;
-            m_gridNodes[secondNode.m_m][secondNode.m_n] = secondNewNodeCoordinates;
+            gridNodes.emplace_back();
         }
-        else
+        gridSizeChanged = true;
+    }
+    if (gridLineType == BoundaryGridLineType::Bottom && areNodesValid)
+    {
+        for (auto& gridNodes : m_gridNodes)
         {
-            m_gridNodes[firstNode.m_m - 1][firstNode.m_n] = firstNewNodeCoordinates;
-            m_gridNodes[secondNode.m_m - 1][secondNode.m_n] = secondNewNodeCoordinates;
+            gridNodes.emplace(gridNodes.begin());
         }
+        gridSizeChanged = true;
     }
-    if (m_gridNodesTypes[firstNode.m_m][firstNode.m_n] == NodeType::Right ||
-        m_gridNodesTypes[secondNode.m_m][secondNode.m_n] == NodeType::Right)
+
+    return gridSizeChanged;
+}
+
+CurvilinearGrid::BoundaryGridLineType CurvilinearGrid::GetBoundaryGridLineType(NodeIndices const& firstNode, NodeIndices const& secondNode) const
+{
```

Comment:
for safety, you could add ta check that the line between firstnnode and secondnode is actually a boundary line. Now it would return that the line is a BoundaryGridLineType::Bottom if firstnode is on the bottom and secondnode is internal, for example